### PR TITLE
Fix initialization of index_array

### PR DESCRIPTION
**Pull request by henry0312**
index_array should be initialized when self.batch_index is zero.

### CODE REVIEW

**Comment by fchollet** [comment]
This will crash if `flow` is called when `self.batch_index` is not 0 (i.e. if you are reusing the same generator more than once, which is common).

**Comment by henry0312** [comment]
> This will crash if `flow` is called when `self.batch_index` is not 0

I don't think so, because:

1. `self.batch_index` is initialized when `flow` is called in the first time.  
    https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L278
2. `index_array` is initialized when `flow` is called in the first time, and `index_array` persists in the scope (i.e. `_flow_index`).  
    https://github.com/fchollet/keras/pull/2590/files#diff-93850fa46a789f2e5905894ad0e7bee4R253
3. `self.batch_index` is incremented as `flow` is called.  
    https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L262
4. If `N < current_index + batch_size` is true, `self.batch_index` will be set 0.  
    https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L265

In other words, `self.batch_index` loops from 0 to `N / batch_size`, and `index_array` is initialized at an appropriate timing (i.e. `index_array` persists in `_flow_index`).

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -249,8 +249,8 @@ def reset(self):
 
     def _flow_index(self, N, batch_size=32, shuffle=False, seed=None):
         while 1:
-            index_array = np.arange(N)
```
Comment (henry0312) [inline keras/preprocessing/image.py]: `index_array` is ALWAYS initialized when `_flow_index`, so some indexes may not be used 😢

**Comment by fchollet** [comment]
No, with the modification you made, if `flow` is called when `self.batch_index` is not `0` (i.e. if `flow` is called twice with the first time only flowing for `n < N`) then `index_array` is not initialized, causing the program to crash. Just try it...

**Comment by MaCaki** [comment]
I agree with @henry0312, the change to the order of initialization of the index_array is causing incorrect behavior: 

```
gen = image.ImageDataGenerator() 
q = gen._flow_index(16, batch_size=5, shuffle=True)  
print(next(q)[0])   
print(next(q)[0])   
print(next(q)[0])   
print(next(q)[0])  
print(next(q)[0])
``` 
produces: 
```
[ 5  0  7 10 14]
[5 6 7 8 9]
[10 11 12 13 14]
[15]
[ 8  3 11  7 15]
```
Note that after the first next() call the array is reinitialized to the un-shuffled range. 
While the previous version was correctly not reinitializing the `index_array` every time the `_flow_index` generator was called by next().

**Comment by henry0312** [comment]
I tried.

## Before

```
» ipython
Python 3.5.1 (default, Feb 13 2016, 10:15:45)
Type "copyright", "credits" or "license" for more information.

IPython 4.2.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import keras
Using Theano backend.

In [2]: from keras.preprocessing.image import ImageDataGenerator

In [3]: gen = ImageDataGenerator()

In [4]: q = gen._flow_index(16, batch_size=5, shuffle=True)

In [5]: print(next(q)[0])
[ 5 14  7  9  0]

In [6]: print(next(q)[0])
[5 6 7 8 9]

In [7]: print(next(q)[0])
[10 11 12 13 14]

In [8]: print(next(q)[0])
[15]

In [9]: print(next(q)[0])
[ 2  9 13  4  5]
```

## After

if this PR is adapted,

```
» ipython
Python 3.5.1 (default, Feb 13 2016, 10:15:45)
Type "copyright", "credits" or "license" for more information.

IPython 4.2.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import keras
Using Theano backend.

In [2]: from keras.preprocessing.image import ImageDataGenerator

In [3]: gen = ImageDataGenerator()

In [4]: q = gen._flow_index(16, batch_size=5, shuffle=True)

In [5]: print(next(q)[0])
[11  1 13  5  2]

In [6]: print(next(q)[0])
[ 6 15  9 10  0]

In [7]: print(next(q)[0])
[12  8  4  3  7]

In [8]: print(next(q)[0])
[14]

In [9]: print(next(q)[0])
[12  2 10  0  9]

In [10]: q = gen._flow_index(16, batch_size=5, shuffle=True)

In [11]: print(next(q)[0])
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-11-2a5d870822b2> in <module>()
----> 1 print(next(q)[0])

/usr/local/var/pyenv/versions/3.5.1/lib/python3.5/site-packages/Keras-1.0.2-py3.5.egg/keras/preprocessing/image.py in _flow_index(self, N, batch_size, shuffle, seed)
    265                 self.batch_index = 0
    266             self.total_batches_seen += 1
--> 267             yield (index_array[current_index: current_index + current_batch_size],
    268                    current_index, current_batch_size)
    269

UnboundLocalError: local variable 'index_array' referenced before assignment
```

---

This PR will fix the initialization of index_array, however, as @fchollet pointed out, it could lead to cause crashes if `flow` is called when `self.batch_index` is not 0
I'll fix it 😀

**Comment by henry0312** [comment]
@MaCaki thanks for your support!

**Comment by henry0312** [comment]
Updated

**Comment by MaCaki** [comment]
If the desired behavior is for the image generator to completely reset ( batch_index=0, index_array reinitialized) each time flow() is called, then could you just do that before the while loop in the generator:
```
    def _flow_index(self, N, batch_size=32, shuffle=False, seed=None):   
      + index_array = np.arange(N)
        while 1:
        ...
```

Since flow() calls reset() before creating another index generator, there doesn't seem to be any issue; 
```
x = np.array([1]*16)
y = np.array([0]*16)
gen.flow(x,y, batch_size=5, shuffle=True)
q = gen.flow_generator 
print(next(q)[0])   
print(next(q)[0]) 

gen.flow(x,y, batch_size=5, shuffle=True)
r = gen.flow_generator 
print(next(r)[0])
print(next(r)[0]) 
```
gives
```
[11  8 12 10  0]
[14 15  4  1 13]
[ 7  8 10  9 14]
[12  5  2  0  6]
```

**Comment by henry0312** [comment]
@MaCaki I think you are almost right (also, I think so. cf. https://github.com/fchollet/keras/pull/2590#issuecomment-216723917), but not all of the users use `flow`. 
Some users may call `_flow_index` directly...

**Comment by MaCaki** [comment]
@henry0312 
Well as a 'weakly internal' function shouldn't some risk be tolerated? 
And even if it's called directly, it won't crash, it just won't reset `batch_index=0`.  If this is the desired behavior, then I see no issue with just doing: 
```
 def _flow_index(self, N, batch_size=32, shuffle=False, seed=None):   
      + index_array = np.arange(N)
      + self.batch_index = 0
         while 1: 
```
or, if someone really wanted to use a `_flow_index()` generator independently of the image generator, they could just call reset() on the `ImageDataGenerator` object first.

**Comment by henry0312** [comment]
okay, updated again.

**Comment by henry0312** [comment]
Added a test case for `_flow_index`

File: tests/keras/preprocessing/test_image.py [hunk 2]
```diff
@@ -90,6 +91,23 @@ def test_img_flip(self):
...
+        generator = ImageDataGenerator()
+        flow_index = generator._flow_index(N, batch_size=batch_size, shuffle=True, seed=seed)
+        for i in range(0, N, batch_size):
+            ary = next(flow_index)[0]
+            if N < i + batch_size:
+                assert generator.batch_index == 0
+            else:
+                assert generator.batch_index == (i // batch_size) + 1
+            assert (ary == idx[i:i+batch_size]).all()
+        assert generator.total_batches_seen == math.ceil(N / batch_size)
```
Comment (fchollet) [inline tests/keras/preprocessing/test_image.py]: Please use `np.ceil` instead

File: tests/keras/preprocessing/test_image.py [hunk 3]
```diff
@@ -90,6 +91,23 @@ def test_img_flip(self):
...
+        N = 11
+        batch_size = 3
+        seed = 0
+        np.random.seed(seed)
+        idx = np.random.permutation(N)
+
+        generator = ImageDataGenerator()
+        flow_index = generator._flow_index(N, batch_size=batch_size, shuffle=True, seed=seed)
+        for i in range(0, N, batch_size):
+            ary = next(flow_index)[0]
```
Comment (fchollet) [inline tests/keras/preprocessing/test_image.py]: Please use readable variable names such as `array`

File: tests/keras/preprocessing/test_image.py (hunk 2 above)
Comment (henry0312) [inline tests/keras/preprocessing/test_image.py]: OK🙆

File: tests/keras/preprocessing/test_image.py (hunk 3 above)
Comment (henry0312) [inline tests/keras/preprocessing/test_image.py]: OK🙆

**Comment by henry0312** [comment]
Updated

**Comment by fchollet** [comment]
The test is incorrectly written and therefore tests are failing.

File: tests/keras/preprocessing/test_image.py [hunk 4]
```diff
@@ -90,6 +90,23 @@ def test_img_flip(self):
...
+
+        generator = ImageDataGenerator()
+        flow_index = generator._flow_index(N, batch_size=batch_size, shuffle=True, seed=seed)
+        for i in range(0, N, batch_size):
+            index_array = next(flow_index)[0]
+            if N < i + batch_size:
+                assert generator.batch_index == 0
+            else:
+                assert generator.batch_index == (i // batch_size) + 1
+            assert (index_array == idx[i:i+batch_size]).all()
```
Comment (MaCaki) [inline tests/keras/preprocessing/test_image.py]: The random permutation `idx` is a completely different random permutation than the one produced internally by the generator.   

Shouldn't you just test that all indices are covered exactly once, i.e. that the union of all `index_arrays` after ciel(N/batch_size) is  [N-1]?

File: keras/preprocessing/image.py [hunk 5]
```diff
@@ -248,9 +248,12 @@ def reset(self):
         self.batch_index = 0
 
     def _flow_index(self, N, batch_size=32, shuffle=False, seed=None):
+        # ensure self.batch_index is 0
+        self.reset()
```
Comment (MaCaki) [inline keras/preprocessing/image.py]: This seems to make `reset()` in `flow()` superfluous.  If so I think it would be cleaner and still safe to just remove the `reset()` from flow and have the index_array generator handle resetting all its counters.

**Comment by fchollet** [comment]
I re-read the PR, and it doesn't make any sense to me.

- the "new" behavior of `flow` is 100% identical to what it is was before.
- the test is testing something that is never meant to be used (a specific private method that is actually a utility function of the public `flow` method, which itself is tested as part of the `ImageDataGenerator` test.

**Comment by MaCaki** [comment]
@fchollet I don't think that the behavior of flow is identical. Here is what's happing under the current master:
```
# current master
x = np.asarray([ [[[i]]*4] for i in range(16)])
y = np.array([1]*16)

gen = image.ImageDataGenerator()
gen.flow(x,y, batch_size=5, shuffle=True)
for i in range(4):
    batch = next(gen)[0].transpose()[0][0]
    print("batch {:d}: {}".format(i, str(batch)))
```
gives
```
batch 0: [[  2.   9.  13.   7.   6.]]
batch 1: [[ 5.  6.  7.  8.  9.]]
batch 2: [[ 10.  11.  12.  13.  14.]]
batch 3: [[ 15.]]
```
I'm printing out the labeled elements of a simpulated batch.  Note that after the first batch **the generator is no longer shuffling**, because the index_array keeps getting reset.  Consequently, the generator will repeat images, or omit them during an epoch of training. Does this test miss something that is happening in practice?  

After simply moving the initialization `index_array = np.arange(N)` to before the while loop, the behavior is correct:
```
batch 0: [[  2.   7.   3.  11.   4.]]
batch 1: [[ 14.  13.  12.  10.   8.]]
batch 2: [[  1.   0.  15.   5.   6.]]
batch 3: [[ 9.]]
```

I fail to see how this is identical.

**Comment by henry0312** [comment]
Are you serious?😲
Please check carefully:

* https://github.com/fchollet/keras/pull/2590#issuecomment-217072816
* https://github.com/fchollet/keras/pull/2590#issuecomment-217292824
* https://github.com/fchollet/keras/pull/2590#discussion_r61984214

**Comment by henry0312** [comment]
https://github.com/fchollet/keras/pull/2590#issuecomment-217279599

> * the test is testing something that is never meant to be used (a specific private method that is actually a utility function of the public flow method, which itself is tested as part of the ImageDataGenerator test.

Do I just remove test codes?

**Comment by fchollet** [comment]
> Do you mean I just remove test codes?

Yes please. I don't think they serve any purpose atm.

**Comment by henry0312** [comment]
I removed test codes.

**Comment by fchollet** [comment]
Thanks, merged.
